# Skills challenge \#5
Below are a series of questions. Use the loaded data to answer the questions. You will almost certainly need to import more packages (`pandas`, `numpy`, etc.) to complete these. You are welcome to use any source except for your classmates. So Google away!

You will be graded on both the **correctness** and **cleanliness** of your work. So don't submit poorly written code or your grade will reflect that. Use Markdown describing what you have done. If you get stuck, move on to another part. Most questions don't rely on the answer to earlier questions.

### Imports

In [1]:
import pandas as pd

### Data loading

In [73]:
df = pd.read_csv('../data/median_rent.csv')

In [74]:
df.head()

City Code          City         Metro        County State  Population Rank  \
0       6181      New York      New York        Queens    NY                1   
1      12447   Los Angeles   Los Angeles   Los Angeles    CA                2   
2      17426       Chicago       Chicago          Cook    IL                3   
3      39051       Houston       Houston        Harris    TX                4   
4      13271  Philadelphia  Philadelphia  Philadelphia    PA                5   

   November 2010  December 2010  January 2011  February 2011  ...  April 2016  \
0            NaN            NaN           NaN            NaN  ...       1.806   
1          1.578          1.578         1.580          1.582  ...       1.990   
2          1.244          1.248         1.254          1.254  ...       1.354   
3          0.788          0.784         0.784          0.786  ...       0.984   
4          0.854          0.858         0.858          0.858  ...       0.948   

   May 2016  June 2016  July 2016  August 2016  September 2016  October 2016  \
0     1.810      1.816      1.824        1.828           1.836         1.844   
1     2.004      2.018      2.026        2.032           2.038         2.042   
2     1.362      1.370      1.374        1.378           1.380         1.380   
3     0.984      0.982      0.980        0.976           0.974         0.974   
4     0.956      0.962      0.964        0.964           0.966         0.968   

   November 2016  December 2016  January 2017  
0          1.858          1.866         1.872  
1          2.048          2.056         2.064  
2          1.380          1.376         1.374  
3          0.976          0.976         0.974  
4          0.972          0.974         0.974  

[5 rows x 81 columns]

### Data description

This data is a collection of median rent per square foot of apartments throughout the US. Each month/year column is the rent per square foot (in USD) in that location.

## Tasks

### Data cleaning
**DC1:** Do any `City Code`s show up more than once?

**DC2:** For each column, print how many `None` or `NaN` values are in the column, along with what percentage of the rows are missing. **Only print columns that actually have missing values**. Round the percentage to two decimal places. Your output should look like:

```
col1_name: 20 (0.05%) missing values 
col3_name: 150 (1.56%) missing values 
```

**DC3:** Drop the `Metro` column, along with `November 2010` and `December 2010`. Save the result back to `df`.

**DC4:** Print how many *rows* have a missing value. Also, print what percentage of the rows have a missing value.

**DC5:** For any rows with missing data, fill the missing values with the median from that row.

### Data exploration

**DE1:** Make a scatterplot and compute the correlation coefficient for `Population Rank` to the median price per square foot in `January 2017`. Write a sentence or two about what we learn from this graph/correlation coefficient.

**DE2:** Find the top ten city codes which had the highest price per square foot in `January 2017`.

### Model building

**MB1:** Split the data into training and testing sets. Your test set should have 30% of the data, and the training set should have the rest.

**MB2:** Create a linear regression model comparing the price per square foot in `January 2011` and the `Population Rank` to the price in `January 2017`. Train it on the training data and test it on the testing data. 

**MB3:** Plot the scatterplot comparing the actual and predicted prices in January 2017. Print the score (correlation coefficient). Label your x and y axes appropriately. All of this should be done on the test set.

**MB4:** If your scatterplot has any strange outliers, explain in plain English what those outliers represent.

**MB5:** Print out the coefficients of the model, along with the y-intercept. Discuss what each of these values tell us. Discuss which variable out of two in your model are more predictive of the target (`January 2017`). Justify your conclusions using your data/model, not just "I feel like...".

### Bonus

**B1:** For each month/year column (say, `January 2012`), make a new column (`January 2012 YOY`) ("YOY" stands for "Year Over Year") which is the percentage increase from the same month in the previous year (so `January 2011` in this case). So for example, if `January 2011` had a value of 1.8 and `January 2012` had a value of 1.85, this is an increase of 0.05, which is a 0.05 / 1.8 = 0.0278 = 2.78% increase. So the column `January 2012 YOY` would have a value of 0.0278. Repeat this for each year (so `January 2013 YOY` would compare to the price in `January 2012`). Don't make a YOY column for the first year (20110. Find which city code has the highest average YOY increase across all years. This sort of computation is incredibly common in financial situations.

**B2:** Explore how it's even possible that a city would have such a large average rent increase.